---
title: "Introduction to `scater`: Single-cell analysis toolkit for expression in R"
author: "Davis McCarthy"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{An introduction to the scater package}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


```{r knitr-options, echo=FALSE}
library(knitr)
opts_chunk$set(fig.align='center', fig.width=6, fig.height=5)
```

This document gives an introduction to and overview of the functionality of the 
`scater` package.

The `scater` package is contains tools to help with the analysis of single-cell 
gene expression data, with the focus on RNA-seq data. The package features:

* A data structure for single-cell expression data;
* Functions for diagnostic plots for quality control;
* Methods for identifying cells at different stages of the cell cycle;
* Differential testing methods for finding genes with differences in behaviour
between experimental conditions.

To get up and running as quickly as possible, see the [Quick Start](#quickstart)
section below. For see the various in-depth sections on various aspects of the 
functionality that follow.

## Quick Start
<a name="quickstart"></a>

Assuming you have a matrix containing expression count data summarised at the 
level of some features (gene, exon, region, etc.), then we first need to form an
`SCESet` object containing the data. An `SCESet` object is the basic data 
container that we use in `scater`. 

Here we use the example data provided with 
the package, which gives us two objects, a matrix of counts and a dataframe 
with information about the cells we are studying:

```{r quickstart-load-data}
library(scater)
data("sc_example_counts")
data("sc_example_cell_info")
ls()
```

We use these objects to form an `SCESet` object containing all of the necessary
information for our analysis:

```{r quickstart-make-sceset}
pd <- new("AnnotatedDataFrame", data = sc_example_cell_info)
rownames(pd) <- pd$Cell
example_sceset <- newSCESet(countData = sc_example_counts, phenoData = pd)
example_sceset
```

Subsetting is very convenient with this class. For example, we can filter out 
genes that are not expressed in any cells:

```{r filter-no-exprs}
keep_gene <- rowSums(is_exprs(example_sceset)) > 0
example_sceset <- example_sceset[keep_gene,]
```

Now we have the expression data neatly stored in a structure that can be used 
for lots of exciting analyses.


## The SCESet class

In `scater` we organise single-cell expression data in objects of the `SCESet` 
class. The class is derived from the Bioconductor `ExpressionSet` class, which 
provides a common interface familiar to those who have analyzed microarray 
experiments with Bioconductor. The class requires three input files:
  1. `exprs`, a numeric matrix of expression values, where rows are 
  genes, and columns are cells
  2. `phenoData`, an `AnnotatedDataFrame` object, where rows are cells, and 
  columns are cell attributes (such as cell type, culture condition, day 
  captured, etc.)
  3. `featureData`, an `AnnotatedDataFrame` object, where rows are features 
  (e.g. genes), and columns are gene attributes, such as biotype, gc content, 
  etc.

The requirements for the `SCESet` class (as with other S4 classes in R and 
Bioconductor) are strict. The idea is that strictness with generating a valid
class object ensures that downstream methods applied to the class will work 
reliably. Thus, the expression value matrix *must* have the same number of 
columns as the `phenoData` object has rows, and it must have the same number of 
rows as the `featureData` dataframe has rows. Row names of the `phenoData` 
object need to match the column names of the expression matrix. Row names of the
`featureData` object need to match row names of the expression matrix.

You can create a new `SCESet` object using count data as follows. In this case, 
the `exprs` slot in the object will be generated as log2(counts-per-million) 
using the `cpm` function from `edgeR`, with a "prior count" value of 1:  

```{r sceset-make-sceset-counts-only}
pd <- new("AnnotatedDataFrame", data = sc_example_cell_info)
rownames(pd) <- pd$Cell
gene_df <- data.frame(Gene = rownames(sc_example_counts))
rownames(gene_df) <- gene_df$Gene
fd <- new("AnnotatedDataFrame", data = gene_df)
example_sceset <- newSCESet(countData = sc_example_counts, phenoData = pd,
                            featureData = fd)
example_sceset
```

We can also make an `SCESet` object with just a matrix of expression values 
(no count data required) like this:

```{r sceset-make-sceset-exprs-only}
example2 <- newSCESet(exprsData = edgeR::cpm.default(sc_example_counts))
pData(example2) 
fData(example2)
```

We have accessor functions to access elements of the `SCESet` object:

* `counts(object)`: returns the matrix of read counts. As you can see above, if 
no counts are defined for the object, then the counts matrix slot is simpy 
`NULL`.
```{r counts-accessor}
counts(example2)[1:3, 1:6]
```
* `exprs(object)`: returns the matrix of gene expression values. Typically these
should be log2(counts-per-million) values or log2(reads-per-kilobase-per-million-mapped),
appropriately normalised of course. The package will generally assume that 
these are the values to use for expression.
```{r exprs-accessor}
exprs(example2)[1:3, 1:6]
```
* `is_exprs(object)`: returns a logical matrix indicating whether each gene 
expression observation is above the defined `lowerDetectionLimit` (default is 0).
This can be determined on the count scale or the "expression" (i.e. `exprs(object)`)
scale.

```{r isexprs-accessor}
is_exprs(example2)[1:3, 1:6]
```

It is straight-forward to change the threshold by which we decide which 
observations are expressed or not:

```{r sceset-define-is-exprs}
is_exprs(example2) <- calcIsExprs(example2, lowerDetectionLimit = 100, 
                               exprs_data = "exprs")
is_exprs(example2)[1:3, 1:6]
```

If you later find some count data that you want to add to the `SCESet` object 
(always worth checking down the back of the couch), then you can add it easily:

```{r sceset-add-count-data}
counts(example2) <- sc_example_counts
example2
counts(example2)[1:3, 1:6]
```

Handily, it is also easy to replace other data in slots of the `SCESet` object 
using generic accessor and replacement functions.

```{r sceset-demo-replacement}
gene_df <- data.frame(Gene = rownames(sc_example_counts))
rownames(gene_df) <- gene_df$Gene
fd <- new("AnnotatedDataFrame", data = gene_df)
## replace featureData
fData(example_sceset) <- fd
## replace phenotype data
pData(example_sceset) <- pd
## replace expression data to be used
exprs(example_sceset) <- edgeR::cpm.default(counts(example_sceset), 
                                            prior.count = 5, log = TRUE)
```


## Plots of expression values

In `scater`, the `plotExpression` function makes it easy to plot expression 
values for a subset of genes or features. This can be particularly useful when 
investigating the some genes identified as being of interest from differential
expression testing or other means.

```{r plot-expression}
plotExpression(example_sceset, rownames(example_sceset)[1:6],
               x = "Mutation_Status", use_as_expr = "exprs")
```

This function uses `ggplot2`, making it easy to change the theme to whatever you 
prefer. We can also show the median expression level per group on the plot and 
show a violin plot to summarise the distribution of expression values:

```{r plot-expression-theme-bw}
plotExpression(example_sceset, rownames(example_sceset)[7:12], 
               x = "Mutation_Status", use_as_expr = "exprs", colour = "Cell_Cycle",
               show_median = TRUE, show_violin = TRUE,  xlab = "Mutation Status")
```


## Quality control

The `scater` package puts a focus on aiding with quality control (QC) and 
pre-processing of single-cell RNA-seq data before further downstream analysis.

We see QC as consisting of three distinct steps:

1. QC and filtering of genes
2. QC and filtering of cells
3. QC of experimental variables

Following QC, we can proceed with data normalisation before downstream analysis 
and modelling.

In the next few sections we discuss the QC and filtering capabilities available
in `scater`.

### Calculate QC metrics

To compute commonly-used QC metrics we have the function `calculateQCMetrics()`:

```{r calc-qc-metrics}
names(pData(example_sceset))
example_sceset <- calculateQCMetrics(example_sceset, gene_controls = 1:20)
varLabels(example_sceset)
```

More than one set of gene controls can be defined if desired.

```{r calc-qc-metrics-multi-gene-controls}
names(pData(example_sceset))
example_sceset <- calculateQCMetrics(
    example_sceset, gene_controls = list(controls1 = 1:20, controls2 = 500:1000),
    cell_controls = list(set_1 = 1:5, set_2 = 31:40))
varLabels(example_sceset)
```

This function adds the following columns to `pData(object)`:

* `depth`: total number of counts for the cell (aka 'library size')
* `log10_depth`: depth on the log10-scale
* `coverage`: the number of genes for the cell that have expression above the 
detection limit (default detection limit is zero)
* `filter_on_depth`: would this cell be filtered out based on its log10-depth 
being (by default) more than 5 median absolute deviations from the median 
log10-depth for the dataset?
* `filter_on_coverage`: would this cell be filtered out based on its coverage 
being (by default) more than 5 median absolute deviations from the median 
coverage for the dataset?
* `counts_from_gene_controls`: total number of counts for the cell that come from 
(a set of user-defined) control genes. Defaults to zero if no control genes are 
indicated.
* `counts_from_endogenous_genes`: total number of counts for the cell that come from 
endogenous genes (i.e. not control genes). Defaults to `depth` if no control genes are 
indicated.
* `log10_counts_from_gene_controls`: total number of counts from control genes on 
the log10-scale. Defaults to zero (i.e. log10(0 + 1), offset to avoid infinite 
values) if no control genes are indicated.
* `log10_counts_from_endogenous_genes`: total number of counts from endogenous 
genes on the log10-scale. Defaults to zero (i.e. log10(0 + 1), offset to avoid 
infinite values) if no control genes are indicated.
* `n_detected_gene_controls`: number of defined gene controls that have 
expression greater than the threshold defined in the object.
*`pct_counts_from_gene_controls`: percentage of all counts that come from the 
defined control genes. Defaults to zero if no control genes are defined.

If we define multiple sets of gene controls, then the above will be supplied for
all gene sets, plus the set of all gene controls combined, as appropriate.

Furthermore, where "counts" appear in the above, the same metrics will also be
computed for "exprs", "tpm" and "fpkm" (if tpm and fpkm are present in the 
`SCESet` object).

The function further adds the following columns to `fData(object)`:

* `mean_exprs`: the mean expression level of the gene/feature.
* `exprs_rank`: the rank of the gene's expression level in the cell.
* `total_gene_counts`: the total number of counts mapped to that gene across all 
cells.
* `log10_total_gene_counts`: total gene counts on the log10-scale.
* `pct_total_counts`: the percentage of all counts that are accounted for by the
counts mapping to the gene.
* `is_gene_control`: is the gene a control gene? Default is `FALSE` unless 
control genes are defined by the user.
`n_cells_exprs`: the number of cells for which the expression level of the gene
is above the detection limit (default detection limit is zero).

### QC and filtering of genes

See `plotFeatureData` below. The subsetting of rows of `SCESet` objects makes it
easy to drop unwanted genes.

### QC and filtering of cells

See `plotPhenoData` and other QC plots below. The subsetting of columns (which 
correspond to cells) of `SCESet` objects makes it easy to drop unwanted cells.

The `plotPCA` function makes it easy to produce a PCA plot directly from an 
`SCESet` object, which is useful for visualising cells.

### QC of experimental variables

See the `plotQC` options below. The various plotting functions enable 
visualisation of the relationship betwen experimental variables and the 
expression data.

### Produce diagnostic plots for QC

Visualising the data and metadata in various ways can be very helpful for QC. We
have a small suite of plotting functions to produce diagnostic plots for:
1. Plotting the most expressed genes across the dataset.
2. Finding the most important principal components for a given cell phenotype or 
metadata variable (from `pData(object)`).
3. Plotting a set of cell phenotype/metadata variables against each other and 
calculating the (marginal) percentage of gene expression variance that they 
explain.

These three QC plots can all be accessed through the function `plotQC` (we need 
to make sure there are no genes with zero or constant expression):

```{r plot-qc}
keep_gene <- rowSums(is_exprs(example_sceset)) > 4
example_sceset <- example_sceset[keep_gene,]
## Plot QC
plotQC(example_sceset, type = "highest-expression")
plotQC(example_sceset, type = "highest-expression", use_as_exprs = "exprs")
plotQC(example_sceset, type = "find-pcs", variable = "coverage")
plotQC(example_sceset, type = "find-pcs", variable = "Cell_Cycle")
plotQC(example_sceset, type = "find-pcs", variable = "Treatment")
plotQC(example_sceset, type = "expl", 
       variables = c("coverage", "depth", "Mutation_Status", "Treatment"))
plotQC(example_sceset, type = "expl")
```

In this small dataset depth and coverage explain a very large proportion of the
variance in gene expression. The proportion of variance that they explain for a 
real dataset should be much smaller (say 1-5%).

We also have neat functions to plot two cell metadata variables:

```{r plot-pdata}
plotPhenoData(example_sceset, aes(x = depth, y = coverage, colour = Mutation_Status))
```

and to plot two feature metadata variables:

```{r plot-fdata}
plotFeatureData(example_sceset, aes(x = n_cells_exprs, y = pct_total_counts))
```

These sorts of plots can be very useful for finding potentially problematic 
cells and genes, as well as identifying important sources of variation in the 
data.

The output of these functions is a `ggplot` object, which can be added to, 
amended and altered. For example, if we don't like the legend position in the 
Coverage vs Depth plot above, we can change it:


```{r plot-pdata-move-legend}
plotPhenoData(example_sceset, aes(x = depth, y = coverage, colour = Mutation_Status)) +
    theme(legend.position = "top")
```

We could also add a trend line for each group:

```{r plot-pdata-move-legend-add-trend}
plotPhenoData(example_sceset, aes(x = depth, y = coverage, colour = Mutation_Status)) +
    theme(legend.position = "top") +
    stat_smooth(method = "lm", se = FALSE, size = 2, fullrange = TRUE)
```

Tapping into the powerful capabilities of `ggplot2`, the possibilities are many.

Combined with the excellent subsetting capabilities of the `SCESet` class, we 
have convenient tools for conducting QC and pre-processing (e.g. filtering) data
for downstream analysis.



## Data normalisation

High levels of variability between cells characterise single-cell expression 
data. In almost all settings, many sources of unwanted variation should be 
accounted for before proceeding with more sophisticated analysis. Below, we show
some of `scater`'s capabilities for normalising data for downstream analyses.

We can use gene controls to help address differences between cells arising from 
different sets of transcripts being expressed and differences in library 
composition.

Important experimental variables and latent factors (if used) can be regressed 
out, so that normalised data has these effects removed.


## Gene and cell pairwise distance matrices

In many single-cell expression analyses we may want to generate and store 
pairwise distance matrices for both cells and genes.

We can first look at a multidimensional scaling (MDS) plot using Euclidean distance
between cells.

```{r cell-pairwise-distance-matrices-euclidean, eval=TRUE}
cell_dist <- as.matrix(dist(t(exprs(example_sceset))))
cellPairwiseDistances(example_sceset) <- cell_dist

if(require(MASS)) {
    fit <- MASS::isoMDS(cellDist(example_sceset), k=2) # k is the number of dim
    # plot solution 
    Col <- c("firebrick", "dodgerblue")[factor(pData(example_sceset)$Treatment)]
    plot(fit$points[,1:2], xlab="Coordinate 1", ylab="Coordinate 2", main="Nonmetric MDS", 
         type="n")
    text(fit$points[,1:2], labels=colnames(example_sceset), col=Col)
}
```

Second, we could also look at an MDS plot using the count data and the Canberra
distance metric.

```{r cell-pairwise-distance-matrices-canberra, eval=TRUE}
cell_dist <- as.matrix(dist(t(counts(example_sceset)), method="canberra"))
cellPairwiseDistances(example_sceset) <- cell_dist

if(require(MASS)) {
    fit <- MASS::isoMDS(cellDist(example_sceset), k=2) # k is the number of dim
    # plot solution 
    Col <- c("firebrick", "dodgerblue")[factor(pData(example_sceset)$Treatment)]
    plot(fit$points[,1:2], xlab="Coordinate 1", ylab="Coordinate 2", main="Nonmetric MDS", 
         type="n")
    text(fit$points[,1:2], labels=colnames(example_sceset), col=Col)
}
```

We can also look at a (nonmetric) MDS plot for the genes, here using Euclidean
distance.

```{r gene-pairwise-distance-matrices, eval=TRUE}
gene_dist <- as.matrix(dist(exprs(example_sceset)))
featurePairwiseDistances(example_sceset) <- gene_dist

if(require(MASS)) {
    fit <- MASS::isoMDS(featDist(example_sceset), k=2) # k is the number of dim
    # plot solution 
    x <- fit$points[,1]
    y <- fit$points[,2]
    plot(x, y, xlab="Coordinate 1", ylab="Coordinate 2", main="Nonmetric MDS", 
         type="n")
    text(x, y, labels = rownames(example_sceset), cex=.7)
}
```


## Using **kallisto** to quantify transcript abundance from within R

Lior Pachter's group at Berkeley has recently released a new software tool 
called **kallisto** for rapid quantification of transcript abundance from RNA-seq
data. In `scater`, a couple of wrapper functions to `kallisto` enable easy and 
extremely fast quantification of transcript abundance from within R.

```{r kallisto-demo-kallisto-test-data, eval=FALSE}
################################################################################
### Tests and Examples

# Example if in the kallisto/test directory
setwd("/home/davis/kallisto/test")
kallisto_log <- runKallisto("targets.txt", "transcripts.idx", single_end=FALSE, 
            output_prefix="output", verbose=TRUE, n_bootstrap_samples=10)

sce_test <- readKallistoResults(kallisto_log, read_h5=TRUE)
sce_test
```

An example using real data from a project investigating cell cycle. Not that 
this analysis is not 'live' (the raw data are not included in the package), but 
it demonstrates what can be done with `scater` and `kallisto`.

```{r kallisto-cell-cycle-example, eval=FALSE}
setwd("/home/davis/021_Cell_Cycle/data/fastq")
system("wc -l targets.txt")
ave_frag_len <- mean(c(855, 860, 810, 760, 600, 690, 770, 690))

kallisto_test <- runKallisto("targets.txt", 
                             "Mus_musculus.GRCm38.rel79.cdna.all.ERCC.idx", 
                             output_prefix="kallisto_output_Mmus", n_cores=12, 
                             fragment_length=ave_frag_len, verbose=TRUE)
sce_kall_mmus <- readKallistoResults(kallisto_test, read_h5=TRUE)
sce_kall_mmus

sce_kall_mmus <- readKallistoResults(kallisto_test)

sce_kall_mmus <- getBMFeatureAnnos(sce_kall_mmus)

head(fData(sce_kall_mmus))
head(pData(sce_kall_mmus))
sce_kall_mmus[["start_time"]]

counts(sce_kall_mmus)[sample(nrow(sce_kall_mmus), size=15), 1:6]

## Summarise expression at the gene level 
sce_kall_mmus_gene <- summariseExprsAcrossFeatures(
    sce_kall_mmus, use_as_exprs="tpm", summarise_by="feature_id")

fData(sce_kall_mmus_gene) %>% datatable

sce_kall_mmus_gene <- getBMFeatureAnnos(
    sce_kall_mmus_gene, filters="ensembl_gene_id", 
    attributes=c("ensembl_gene_id", "mgi_symbol", "chromosome_name", 
                 "gene_biotype", "start_position", "end_position",
                 "percentage_gc_content", "description"), 
    feature_symbol="mgi_symbol", feature_id="ensembl_gene_id",
    biomart="ensembl", dataset="mmusculus_gene_ensembl")

fData(sce_kall_mmus_gene) %>% datatable

## Add gene symbols to featureNames to make them more intuitive
new_feature_names <- featureNames(sce_kall_mmus_gene)
notna_mgi_symb <- !is.na(fData(sce_kall_mmus_gene)$mgi_symbol)
new_feature_names[notna_mgi_symb] <- fData(sce_kall_mmus_gene)$mgi_symbol[notna_mgi_symb]
notna_ens_gid <- !is.na(fData(sce_kall_mmus_gene)$ensembl_gene_id)
new_feature_names[notna_ens_gid] <- paste(new_feature_names[notna_ens_gid],
          fData(sce_kall_mmus_gene)$ensembl_gene_id[notna_ens_gid], sep="_")
sum(duplicated(new_feature_names))
featureNames(sce_kall_mmus_gene) <- new_feature_names
featureNames(sce_kall_mmus_gene) %>% head
featureNames(sce_kall_mmus_gene) %>% tail
sum(duplicated(fData(sce_kall_mmus_gene)$mgi_symbol))

```





